chain/mempool_test.go

```diff
@@ -81,13 +128,76 @@ func TestCachedInputsAddInput(t *testing.T) {
 	require.True(ok)
 	require.Equal(tx.TxHash(), txid)
 
+	// Check the internal state. Since we've just added one input, there
+	// should be exactly one item in each map.
+	require.Len(c.inputs, 1)
+	require.Len(c.txids, 1)
+	require.Len(c.txids[tx.TxHash()], 1)
+
 	// Add the input again using the replacement tx.
 	c.addInput(op, replacedTx.TxHash())
 
 	// Lookup should now give us the replacement txid.
 	txid, ok = c.hasInput(op)
 	require.True(ok)
 	require.Equal(replacedTx.TxHash(), txid)
+
+	// Check the internal state.
+	//
+	// We should only have one input.
+	require.Len(c.inputs, 1)
+
+	// Expect two transactions.
+	require.Len(c.txids, 2)
+
+	// The new txid should be present.
+	require.Len(c.txids[replacedTx.TxHash()], 1)
+
+	// The old txid should be an empty map.
+	require.Empty(c.txids[tx.TxHash()])
+}
+
+// TestCachedInputsAddInputTwice checks that when the same tx is added again it
+// won't change the cachedInputs's internal state.
+func TestCachedInputsAddInputTwice(t *testing.T) {
```

Comment:
Confirmed this test failed if the above diff is modified to not use `isReplacement`:
```
=== RUN   TestCachedInputsAddInputTwice
    mempool_test.go:202:
        	Error Trace:	/Users/roasbeef/gocode/src/github.com/btcsuite/btcwallet/chain/mempool_test.go:202
        	Error:      	"map[]" should have 1 item(s), but has 0
        	Test:       	TestCachedInputsAddInputTwice
```

I would've thought `TestCachedInputsAddInputReplacement` would fail as well though, as iiuc, the issue was that we weren't also removing the _old_ now replaced `txid`.

Reply:
fwiw the issue was when adding the same tx multiple times, we'd delete it every second time we added it.